### PR TITLE
Sighash docs

### PR DESCRIPTION
**Pull request by ErinHales**
- Adds docs to `sighash.js`
 - Updates documentation in `ValidateSPV.js`, using typedefs is so much cleaner

### CODE REVIEW

File: js/src/sighash.js [hunk 1]
```diff
@@ -6,12 +6,51 @@ import * as BTCUtils from './BTCUtils';
...
+ * @property {string}     vin The vin
+ * @property {string}     vout The vout
+ * @property {string}     locktime 4-byte locktime
+ */
+
+/**
+ * 
+ * Validates a flag
+ * 
+ * @param {number}      flag The first byte of a VarInt
```
Comment (prestwich) [inline js/src/sighash.js:38]: this is a sighash flag. different flag from varint flags

File: js/src/sighash.js [hunk 2]
```diff
@@ -43,7 +99,17 @@ export function hashPrevouts(inputs, flag) {
...
-// inputs is an array of Uint8Arrays
+/**
+ *
+ * Hashes sequence according to BIP143 semantics.
+ * 
+ * For BIP143 (Witness and Compatibility sighash) documentation, see here:
+ * - https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+ *
+ * @param {array}         inputs An array of inputs (type Uint8Array)
+ * @param {number}        flag The first byte of a VarInt
```
Comment (prestwich) [inline js/src/sighash.js:109]: sighash flag, not varint flag

File: js/src/sighash.js [hunk 3]
```diff
@@ -43,7 +99,17 @@ export function hashPrevouts(inputs, flag) {
...
+ * @returns {Uint8Array}  The hash of the Sequence
```
Comment (prestwich) [inline js/src/sighash.js:110]: `BIP143 hashSequence`

File: js/src/sighash.js [hunk 4]
```diff
@@ -54,15 +120,32 @@ export function hashSequence(inputs, flag) {
...
-// outputs is an array of Uint8Arrays
+/**
+ *
+ * Hashes outputs according to BIP143 semantics.
+ * 
+ * For BIP143 (Witness and Compatibility sighash) documentation, see here:
+ * - https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+ *
+ * @param {array}         outputs An array of outputs (type Uint8Array)
+ * @returns {Uint8Array}  The hash of the Outputs
```
Comment (prestwich) [inline js/src/sighash.js:130]: `BIP143 hashOutputs`

File: js/src/sighash.js [hunk 5]
```diff
@@ -54,15 +120,32 @@ export function hashSequence(inputs, flag) {
...
 }
 
-// Check if nSequence locks might be active
+/**
+ *
+ * Checks if nSequence locks might be active
+ *
+ * @param {array}         inputs An array of inputs (type Uint8Array)
+ * @param {Uint8Array}    locktime 4-byte tx locktime
+ * @param {number}        flag The first byte of a VarInt
```
Comment (prestwich) [inline js/src/sighash.js:145]: sighash flag, not varint flag

File: js/src/sighash.js [hunk 6]
```diff
@@ -88,7 +178,20 @@ export function possibleRelativeLock(inputs, version) {
...
+/**
+ *
+ * Calculates sighash
+ * 
+ * @dev All args are deserialized
+ *
+ * @param {tx}            tx The tx
+ * @param {number}        index The index
+ * @param {number}        sighashFlag The first byte of a VarInt
+ * @param {}              prevoutScript
```
Comment (prestwich) [inline js/src/sighash.js:192]: uint8array

File: js/src/sighash.js [hunk 7]
```diff
@@ -88,7 +178,20 @@ export function possibleRelativeLock(inputs, version) {
...
+ * @param {}              prevoutValue
```
Comment (prestwich) [inline js/src/sighash.js:193]: uint8array

File: js/src/sighash.js [hunk 8]
```diff
@@ -88,7 +178,20 @@ export function possibleRelativeLock(inputs, version) {
...
+// TODO: make typedef for tx
+/**
+ *
+ * Calculates sighash
+ * 
+ * @dev All args are deserialized
+ *
+ * @param {tx}            tx The tx
+ * @param {number}        index The index
+ * @param {number}        sighashFlag The first byte of a VarInt
```
Comment (prestwich) [inline js/src/sighash.js:191]: sighash flag, not varint flag

File: js/src/sighash.js [hunk 9]
```diff
@@ -130,8 +233,16 @@ export function sighash(tx, index, sighashFlag, prevoutScript, prevoutValue) {
...
 
-// deserializes the args from hex
-// serTx is { version, vin, vout, locktime } all as hex
+/**
+ *
+ * Deserializes the args for `sighash` from hex
+ *
+ * @param {SerTx}         serTx The tx all as hex
+ * @param {}              serPrevoutScript
+ * @param {}              serPrevoutValue
```
Comment (prestwich) [inline js/src/sighash.js:248]: both of these are uint8arrays

File: js/src/sighash.js [hunk 10]
```diff
@@ -144,8 +255,18 @@ export function deserSighashArgs(serTx, serPrevoutScript, serPrevoutValue) {
...
 
-// runs `deserSighashArgs` and then `sighash`
-export function deserAndSighash(tx, index, sighashFlag, prevoutScript, prevoutValue) {
-  const deser = deserSighashArgs(tx, prevoutScript, prevoutValue);
+/**
+ *
+ * Runs `deserSighashArgs` and then`sighash`
+ *
+ * @param {SerTx}         serTx The tx all as hex
+ * @param {}              index
```
Comment (prestwich) [inline js/src/sighash.js:269]: number

File: js/src/sighash.js [hunk 11]
```diff
@@ -144,8 +255,18 @@ export function deserSighashArgs(serTx, serPrevoutScript, serPrevoutValue) {
...
+ * @param {number}        sighashFlag The first byte of a VarInt
+ * @param {}              prevoutScript
+ * @param {}              prevoutValue
```
Comment (prestwich) [inline js/src/sighash.js:272]: both of these are hex strings

**Comment by prestwich** [comment]
where did the golang file changes come from? :)

**Comment by ErinHales** [comment]
I'm... not sure.  That's weird, one sec. I just merged your branch into mine...

**Comment by ErinHales** [comment]
Those are changes that are on master. I'm not sure how they got on this branch though...

**Comment by prestwich** [comment]
okay one sec

**Pull request merged**